components/sup/src/event.rs

```diff
@@ -109,9 +97,49 @@ pub struct EventCore {
     pub supervisor_id: String,
 }
 
+/// Send an event for the start of a Service.
+pub fn service_started(service: &Service) {
+    if stream_initialized() {
+        publish(ServiceStartedEvent { service_metadata:    Some(service.to_service_metadata()),
+                                      supervisor_metadata: None, });
+    }
+}
+
+/// Send an event for the stop of a Service.
+pub fn service_stopped(service: &Service) {
+    if stream_initialized() {
```

Comment:
Are there going to be more events in the future that use this same pattern? If so, perhaps moving the `if stream_initialized()` check inside `publish` would reduce duplication.

Or, to reduce the duplication of `EVENT_STREAM.try_get`, `stream_initialized` could be changed to something like `try_get_event_stream`, and these calls would become
```rust
    if let Some(event_stream) = try_get_event_stream() {
        publish(event_stream, …);
```

Reply:
I agree that the way the code is now, it's more repetitive than I'd like. My goal (which may admittedly be in the premature optimization camp) was to not do anything extra if you don't have the event stream enabled. By placing the initialization check into the event functions themselves, rather than down in the `publish` function, we would never have to make copies and allocate a protobuf object that we're just going to throw away.

I do like the second suggestion, though.

Reply:
Actually, I'd like to hold off until we have a few more events in play. I think the `try_get_event_stream` actually leads to a bit more duplication.